dapps/valuetransfers/packages/tangle/signature_filter.go

```diff
@@ -0,0 +1,90 @@
+package tangle
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/iotaledger/hive.go/async"
+	"github.com/iotaledger/hive.go/autopeering/peer"
+
+	"github.com/iotaledger/goshimmer/dapps/valuetransfers/packages/payload"
+	"github.com/iotaledger/goshimmer/packages/binary/messagelayer/message"
+	"github.com/iotaledger/goshimmer/packages/binary/messagelayer/messageparser"
+)
+
+// SignatureFilter represents a filter for the MessageParser that filters out transactions with an invalid signature.
+type SignatureFilter struct {
+	onAcceptCallback func(message *message.Message, peer *peer.Peer)
+	onRejectCallback func(message *message.Message, err error, peer *peer.Peer)
+	workerPool       async.WorkerPool
+
+	onAcceptCallbackMutex sync.RWMutex
+	onRejectCallbackMutex sync.RWMutex
+}
+
+// NewSignatureFilter is the constructor of the MessageFilter.
+func NewSignatureFilter() (result *SignatureFilter) {
+	result = &SignatureFilter{}
+
+	return
+}
+
+// Filter get's called whenever a new message is received. It first checks if the message contains a value Payload and
+// then verifies the signature.
+func (filter *SignatureFilter) Filter(message *message.Message, peer *peer.Peer) {
+	filter.workerPool.Submit(func() {
+		if valuePayload := message.Payload(); valuePayload.Type() == payload.Type {
```

Comment:
I think this can be more go idiomatic and much simpler to read from top to bottom if you don't nest the if statements but simply return with an error message. If all checks are passed `filter.getAcceptCallback()(message, peer)` is called at the very end of the function.